memcached-operator/pkg/controller/memcached/memcached_controller.go

```diff
@@ -109,55 +109,56 @@ func (r *ReconcileMemcached) Reconcile(request reconcile.Request) (reconcile.Res
 		return reconcile.Result{}, err
 	}
 
-	// Check if the Deployment already exists, if not create a new one
-	deployment := &appsv1.Deployment{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Name: memcached.Name, Namespace: memcached.Namespace}, deployment)
-	if err != nil && errors.IsNotFound(err) {
-		// Define a new Deployment
-		dep := r.deploymentForMemcached(memcached)
-		reqLogger.Info("Creating a new Deployment.", "Deployment.Namespace", dep.Namespace, "Deployment.Name", dep.Name)
-		err = r.client.Create(context.TODO(), dep)
-		if err != nil {
-			reqLogger.Error(err, "Failed to create new Deployment.", "Deployment.Namespace", dep.Namespace, "Deployment.Name", dep.Name)
-			return reconcile.Result{}, err
-		}
-		// Deployment created successfully - return and requeue
-		// NOTE: that the requeue is made with the purpose to provide the deployment object for the next step to ensure the deployment size is the same as the spec.
-		// Also, you could GET the deployment object again instead of requeue if you wish. See more over it here: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/reconcile#Reconciler
-		return reconcile.Result{Requeue: true}, nil
-	} else if err != nil {
-		reqLogger.Error(err, "Failed to get Deployment.")
-		return reconcile.Result{}, err
-	}
+	// Update Deployment if it has changed to something other than expected. If the Deployment
+	// doesn't exist it will be created (initial creation is not really a special case).
+	deployment := &appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{Name: memcached.Name, Namespace: memcached.Namespace}}
+	op, err := controllerutil.CreateOrUpdate(context.TODO(), r.client, deployment, func(ignore runtime.Object) error {
+
+		// Set deployment object to expected state in memory. If it's different than what
+		// the calling `CreateOrUpdate` function gets from Kubernetes, it will send an
+		// update request back to the apiserver with the expected state determined here.
+		deploymentForMemcached(memcached, deployment)
 
-	// Ensure the deployment size is the same as the spec
-	size := memcached.Spec.Size
-	if *deployment.Spec.Replicas != size {
-		deployment.Spec.Replicas = &size
-		err = r.client.Update(context.TODO(), deployment)
+		// Set Memcached instance as the owner of the Deployment.
+		err := controllerutil.SetControllerReference(memcached, deployment, r.scheme)
 		if err != nil {
-			reqLogger.Error(err, "Failed to update Deployment.", "Deployment.Namespace", deployment.Namespace, "Deployment.Name", deployment.Name)
-			return reconcile.Result{}, err
+			reqLogger.Error(err, "Failed to set owner reference on memcached Deployment")
```

Comment:
I would suggest keeping the logs as before since is very useful to debug the app knows the namespace and resource name. It could help the users with they decide to use it in their POCs tests as well.

Reply:
We should keep the `Deployment.Namespace` and `Deployment.Name` fields in the log, but the message itself is good since this error is about setting the owner reference.